File: client/client-multi/client-objective-c/Kaa/channel/impl/transports/DefaultProfileTransport.m
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "DefaultProfileTransport.h"
#import "KeyUtils.h"
#import "KaaLogging.h"

#define TAG @"DefaultProfileTransport >>>"

@interface DefaultProfileTransport ()

@property (nonatomic, strong) id<ProfileManager> profileMgr;
@property (nonatomic, strong) KaaClientProperties *properties;

- (BOOL)isProfileOutdated:(EndpointObjectHash *)currentProfileHash;

@end

@implementation DefaultProfileTransport

- (void)sync {
    [self syncAll:TRANSPORT_TYPE_PROFILE];
}

- (ProfileSyncRequest *)createProfileRequest {
    if (self.clientState && self.profileMgr && self.properties) {
        NSData *serializedProfile = [self.profileMgr getSerializedProfile];
        EndpointObjectHash *currentProfileHash = [EndpointObjectHash hashWithSHA1:serializedProfile];
        if ([self isProfileOutdated:currentProfileHash]
            || ![self.clientState isRegistred]
            || [self.clientState needProfileResync]) {
            [self.clientState setProfileHash:currentProfileHash];
            ProfileSyncRequest *request = [[ProfileSyncRequest alloc] init];
            request.endpointAccessToken = [KAAUnion unionWithBranch:KAA_UNION_STRING_OR_NULL_BRANCH_0
                                                               data:[self.clientState endpointAccessToken]];
            if (![self.clientState isRegistred]) {
                request.endpointPublicKey = [KAAUnion unionWithBranch:KAA_UNION_BYTES_OR_NULL_BRANCH_0
                                                                 data:[self.clientState publicKeyAsBytes]];
            }
            request.profileBody = serializedProfile;
            return request;
            
        } else {
            DDLogInfo(@"%@ Profile is up to date", TAG);
        }
    } else {
        DDLogError(@"%@ Failed to create ProfileSyncRequest clientState %@, manager %@, properties %@",
                   TAG, self.clientState, self.profileMgr, self.properties);
    }
    return nil;
}

- (void)onProfileResponse:(ProfileSyncResponse *)response {
    if (response.responseStatus == SYNC_RESPONSE_STATUS_RESYNC) {
        [self.clientState setNeedProfileResync:YES];
        [self syncAll:[self getTransportType]];
    } else if (self.clientState && ![self.clientState isRegistred]) {
        [self.clientState setIsRegistred:YES];
    }
    DDLogInfo(@"%@ Processed profile response", TAG);
}

- (void)setProfileManager:(id<ProfileManager>)manager {
    self.profileMgr = manager;
}

- (void)setClientProperties:(KaaClientProperties *)clientProperties {
    self.properties = clientProperties;
}

- (TransportType)getTransportType {
    return TRANSPORT_TYPE_PROFILE;
}

- (BOOL)isProfileOutdated:(EndpointObjectHash *)currentProfileHash {
    EndpointObjectHash *currentHash = [self.clientState profileHash];
    return !currentHash || ![currentProfileHash isEqual:currentHash];
}

@end

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/channel/impl/transports/DefaultProfileTransport.java
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.kaaproject.kaa.client.channel.impl.transports;

import org.kaaproject.kaa.client.KaaClientProperties;
import org.kaaproject.kaa.client.channel.ProfileTransport;
import org.kaaproject.kaa.client.profile.ProfileManager;
import org.kaaproject.kaa.common.TransportType;
import org.kaaproject.kaa.common.endpoint.gen.ProfileSyncRequest;
import org.kaaproject.kaa.common.endpoint.gen.ProfileSyncResponse;
import org.kaaproject.kaa.common.endpoint.gen.SyncResponseStatus;
import org.kaaproject.kaa.common.hash.EndpointObjectHash;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.nio.ByteBuffer;

public class DefaultProfileTransport extends AbstractKaaTransport implements
    ProfileTransport {

  private static final Logger LOG = LoggerFactory.getLogger(DefaultProfileTransport.class);

  private ProfileManager manager;
  private KaaClientProperties properties;

  private boolean isProfileOutDated(EndpointObjectHash currentProfileHash) {
    EndpointObjectHash currentHash = clientState.getProfileHash();
    return currentHash == null || !currentHash.equals(currentProfileHash);
  }

  @Override
  public void sync() {
    syncAll(TransportType.PROFILE);
  }

  @Override
  public ProfileSyncRequest createProfileRequest() throws IOException {
    if (clientState != null && manager != null && properties != null) {
      byte[] serializedProfile = manager.getSerializedProfile();
      EndpointObjectHash currentProfileHash = EndpointObjectHash.fromSha1(serializedProfile);
      if (isProfileOutDated(currentProfileHash)
          || !clientState.isRegistered()
          || clientState.isNeedProfileResync()) {
        clientState.setProfileHash(currentProfileHash);
        ProfileSyncRequest request = new ProfileSyncRequest();
        request.setEndpointAccessToken(clientState.getEndpointAccessToken());
        if (!clientState.isRegistered()) {
          request.setEndpointPublicKey(ByteBuffer.wrap(clientState.getPublicKey().getEncoded()));
        }
        request.setProfileBody(ByteBuffer.wrap(serializedProfile));
        return request;
      } else {
        LOG.info("Profile is up to date");
      }
    } else {
      LOG.error("Failed to create ProfileSyncRequest clientState {}, manager {}, properties {}",
              clientState, manager, properties);
    }
    return null;
  }

  @Override
  public void onProfileResponse(ProfileSyncResponse response) throws Exception {
    if (response.getResponseStatus() == SyncResponseStatus.RESYNC) {
      clientState.setIfNeedProfileResync(true);
      sync();
    } else if (clientState != null && !clientState.isRegistered()) {
      clientState.setRegistered(true);
    }
    LOG.info("Processed profile response");
  }

  @Override
  public void setProfileManager(ProfileManager manager) {
    this.manager = manager;
  }

  @Override
  public void setClientProperties(KaaClientProperties properties) {
    this.properties = properties;
  }

  @Override
  protected TransportType getTransportType() {
    return TransportType.PROFILE;
  }

}
